tracer/src/Datadog.Trace/Activity/Handlers/DefaultActivityHandler.cs

```diff
@@ -84,16 +95,23 @@ public void ActivityStarted<T>(string sourceName, T activity)
                     return;
                 }
 
-                ActivityScope.GetOrAdd(activity.Instance, _ => CreateScopeFromActivity(activity, traceId, spanId, rawTraceId, rawSpanId));
+                ActivityMappingById.GetOrAdd(activity.Id, _ => new(activity.Instance, CreateScopeFromActivity(activity, parent, traceId, spanId, rawTraceId, rawSpanId)));
             }
             catch (Exception ex)
             {
                 Log.Error(ex, "Error processing the OnActivityStarted callback");
             }
 
-            static Scope CreateScopeFromActivity(T activity, ulong? traceId, ulong? spanId, string? rawTraceId, string? rawSpanId)
+            static Scope CreateScopeFromActivity(T activity, SpanContext? parent, ulong? traceId, ulong? spanId, string? rawTraceId, string? rawSpanId)
             {
-                var span = Tracer.Instance.StartSpan(activity.OperationName, startTime: activity.StartTimeUtc, traceId: traceId, spanId: spanId, rawTraceId: rawTraceId, rawSpanId: rawSpanId);
+                string? serviceName = activity switch
+                {
+                    IActivity5 activity5 => activity5.Source.Name,
+                    _ => null
+                };
+
+                var span = Tracer.Instance.StartSpan(activity.OperationName, parent: parent, serviceName: serviceName, startTime: activity.StartTimeUtc, traceId: traceId, spanId: spanId, rawTraceId: rawTraceId, rawSpanId: rawSpanId);
```

Comment:
Not specific to this PR, but just wondering: currently we're passing in a null `Tags` object, so we create a `CommonTags` object. Given we add a _bunch_ of tags later on, is it worth using an `OtelTags` (or similar) object, so we can use properties and avoid all the dictionary look ups? Could/should be a separate PR probably if so.

Reply:
Yep this is a good idea. I didn't start with that optimization because I was still figuring out the set of tags needed, but it makes sense now to create a new Tags implementation with _at least_ the set of tags specified in the span metadata rules. Let's do that separately 👍🏼